File: scripts/prediction/plot_prediction_accuracy.py
"""
Plot accuracy output from binary classification
as a bar graph of averages +/- standard deviations.
"""
from __future__ import division
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from argparse import ArgumentParser
import os

def main():
    parser = ArgumentParser()
    parser.add_argument('accuracy_scores')
    parser.add_argument('--out_dir', default='../../output')
    args = parser.parse_args()
    accuracy_file = args.accuracy_scores
    out_dir = args.out_dir
    out_file = accuracy_file.replace('.tsv', '.png')
    
    # load data, aggregate
    accuracy_scores = pd.read_csv(accuracy_file, sep='\t', index_col=0)
    folds = accuracy_scores.loc[:, 'folds'].unique()[0]
    accuracy_scores.loc[:, 'AccuracySD'] = accuracy_scores.loc[:, 'AccuracySD'] * 100 / (folds)**.5
    accuracy_scores.loc[:, 'Accuracy'] = accuracy_scores.loc[:, 'Accuracy'] * 100
    
    # aggregate and sort by accuracy
    accuracy_means = accuracy_scores.loc[:, ['Accuracy', 'AccuracySD', 'feature_set_name']].groupby(by='feature_set_name').apply(pd.np.mean)
    accuracy_means = accuracy_means.sort_values('Accuracy', inplace=False, ascending=True)
    
    # plot!
    # TACL size
#     tick_size = 14
    # NWAV size
    tick_size = 20
    # TACL size
#     label_size = 20
    # NWAV size
    label_size = 28
    annotate_font_size = 12
    xlabel = 'Feature set'
    ylabel = 'Accuracy'
    # bar_color = 'b'
    bar_color = (117, 117, 255) # light blue
    bar_color = (c/255 for c in bar_color)
    bar_edge_color = 'k'
    error_color = 'k'
    feature_set_names = accuracy_means.index.tolist()
    feature_set_names_script = map(lambda n: '$\mathtt{%s}$'%(n), feature_set_names)
    feature_set_count = len(feature_set_names)
    bar_x = pd.np.arange(1, feature_set_count+1)
    bar_y = accuracy_means.loc[:, 'Accuracy']
    y_err = accuracy_means.loc[:, 'AccuracySD']
    x_offset = 0.18
    y_offset = 2.2
    error_capsize = 5
    annotate_x = bar_x - x_offset
    annotate_y = bar_y + y_offset
    plt.bar(bar_x, bar_y, color='None', edgecolor=bar_edge_color)
    # add error bars
    plt.errorbar(bar_x, bar_y, yerr=y_err, fmt='none', ecolor=error_color, capsize=error_capsize)
    # add mean annotation
    for x,y,accuracy in zip(annotate_x, annotate_y, bar_y):
        accuracy_str = '%.2f'%(accuracy)
        plt.annotate(accuracy_str, xy=(x,y), xytext=(x,y), fontsize=annotate_font_size)
    # add labels
    plt.xticks(bar_x, feature_set_names_script, fontsize=tick_size)
    plt.yticks(fontsize=tick_size)
    plt.xlabel(xlabel, fontsize=label_size)
    plt.ylabel(ylabel, fontsize=label_size)
    plt.tight_layout()
    plt.savefig(out_file)
    
if __name__ == '__main__':
    main()

File: scripts/prediction/plot_prediction_accuracy.R
# make bar plot of success/failure prediction accuracy
library('ggplot2')

args <- commandArgs(trailingOnly = TRUE)
accuracy_file <- args[1]
out_dir <- args[2]
file_base <- basename(accuracy_file)
out_file <- file.path(out_dir, gsub('.tsv', '.png', file_base))
print(out_file)

# load data, aggregate
print(accuracy_file)
accuracy_results <- read.table(accuracy_file, sep='\t', header = TRUE, check.names = FALSE)
print(head(accuracy_results))
accuracy_results[, 'feature_set_name'] <- as.factor(accuracy_results[, 'feature_set_name'])
folds <- unique(accuracy_results[, 'folds'])[1]
accuracy_results[, 'AccuracySD'] <- accuracy_results[, 'AccuracySD'] * 100 / (folds)**.5
accuracy_results[, 'Accuracy'] <- accuracy_results[, 'Accuracy'] * 100

accuracy_aggregate <- aggregate(as.formula('. ~ feature_set_name'), accuracy_results, mean)
# sort by accuracy
accuracy_aggregate <- accuracy_aggregate[with(accuracy_aggregate, order(Accuracy)), ]

## plot!!
barplot_margins = c(5, 5, 5, 5)
par(mar = barplot_margins)
bar_x <- 1:(dim(accuracy_aggregate)[1])
bar_y <- accuracy_aggregate[, 'Accuracy']
err_y <- accuracy_aggregate[, 'AccuracySD']
feature_set_names <- accuracy_aggregate[, 'feature_set_name']
ylabel <- 'Accuracy'
# ylim <- c(min(bar_y) - max(err_y)*2, max(bar_y) + max(err_y)*2)
ylim <- c(0, max(bar_y) + max(err_y)*4)
png(filename=out_file)
bars <- barplot(height = bar_y, ylim=ylim, names.arg = feature_set_names, ylab = ylabel)
err_bars <- segments(bars, bar_y - err_y, bars, bar_y + err_y, lwd = 1.5)
err_arrows <- arrows(bars, bar_y - err_y, bars, bar_y + err_y, lwd = 1.5, angle = 90, code = 3, length = 0.05)

# add annotations with mean values
max_digits <- 4
annotate_y_offset <- 2.
annotate_cex <- 2.0
mean_str <- lapply(bar_y, function(x){return(format(x, digits=max_digits))})
annotate_y <- bar_y + annotate_y_offset
text(bars, y=annotate_y, labels=mean_str, cex=annotate_cex)
dev.off()